### DISCUSSION TITLE
Questions about Downloading and Compiling the Code

### DISCUSSION THREAD

**Discussion by murphy-john** [Anura3D/Anura3D_OpenSource#3] [General]
Please add your questions about downloading and compiling the code here.

**Comment by yuguzhangwu** [comment]
Hi everyone，
    I'm a PhD student in Geotechnics. I have downloaded the Codes and tried to compile them. But there are lots of warnings(981 warnings) and I think I failed because it doesn't work. Would you please help me to deal with the compiling part? Here are part of my error list. Thanks a lot!
![b5a9fda85d8946118b141ae6424fc46](https://user-images.githubusercontent.com/106877284/172042160-7a033193-6a8d-4a42-bf9e-edb28e0b09ef.jpg)
![5321a69a90b567f55031a5567e5d918](https://user-images.githubusercontent.com/106877284/172042161-5528b1e8-53de-4045-8dfb-2cd5c17f44e2.jpg)
![79ce1f2ff9a984b6b1579d337249d0c](https://user-images.githubusercontent.com/106877284/172042162-57a476bc-17e5-4ff0-9de2-6b73181059a9.jpg)

**Reply by murphy-john** [reply]
Hello! Will you share a picture of the bottom of the compiling messages? Does it look like this photo? Does it say "Build: 4 Succeeded..."?
<img width="385" alt="Capture" src="https://user-images.githubusercontent.com/47335381/172060786-deea7d36-0e70-4d6c-a26e-a262728db865.PNG">
Warnings aren't usually a big deal when compiling the Anura3D source code. Most of the warnings you see when compiling in Visual Studio are just letting us know we could to a better job cleaning up some variable definitions and removing unused local variables. 

If you see a red exclamation point icon and "Error..." that's a sign the code didn't compile correctly. 

When you say "...it didn't work.", what is the "it" that isn't working? Is it an issue with compiling or something with running a problem with the code? Have you looked at the tutorial manual and/or YouTube videos?

**Reply by yuguzhangwu** [reply]
Hello! Thank you very much for replying me. I have got the the picture you share. And it said  "Build: 4 Succeeded..." Also, there is no error when I was compiling the code. But then, I tried to run a problem followed the YouTube video and also looked at the tutorial manual. After finishing the work in GiD, I double clicked calculate.bat in the first picture and the it doesn't work(picture 2). So, I wonder if there is something wrong when compiling the code. 
![微信图片_202206062332011](https://user-images.githubusercontent.com/106877284/172193672-f62c9e2c-7c30-420b-989e-ae61ddc981ae.jpg)
![微信图片_20220606233201](https://user-images.githubusercontent.com/106877284/172193692-771376ee-28d9-403b-bd54-b7f92fa6fa08.jpg)

**Reply by murphy-john** [reply]
Ah! Sorry, this is a known bug where the soil model DLL files are put in a sub-folder instead of in the model (*.A3D) folder. I think there is a simple work around, copy the file called "A3DLinearElasticity.dll" from the "SoilModels" folder and paste the files into the "oedometer.A3D" folder. Then try running the code. You'll have to do that for any of the soil model DLL files you use. 

You can check the list of files needed to perform the oedometer tutorial example in Section 3.3 of the tutorial manual.

**Reply by yuguzhangwu** [reply]
Wow! Thanks a lot! This error has disappeared. But new errors appeared showed in the following picture.
![1654562274020](https://user-images.githubusercontent.com/106877284/172273775-1bef4db5-b7bd-46fd-b572-59dbf716dd66.jpg)

**Reply by yuguzhangwu** [reply]
> 

Ah, I found I made a silly mistake that I copied the wrong Anura3D_2022.exe. Now it works. Thanks a lot for your help!

**Comment by kuailebozai** [comment]
Hello, I'm also learning to use anura3d, but I can't download anura3d_ 2022.exe. If you have the latest version, please tell me how to download or send the file to me. 
My email address is: kuailebozai@stu.csust.edu.cn
Thanks a lot for your help!

**Reply by ayerrocolom** [reply]
Hello,
The executable cannot be downloaded. You have to compile the code following the instructions detailed in the [tutorial manual](https://github.com/Anura3D/Anura3D_OpenSource/blob/main/manuals/TutorialManual_2022.pdf).

**Reply by kuailebozai** [reply]
Ah! Thanks a lot for your help! I have to compile the code myself.

**Reply by ayerrocolom** [reply]
Yes! :)

**Reply by kuailebozai** [reply]
Hi, I'm trying to compile the code, but I can't find the "DynamicMPM.sln" file in the code base. What should I do? Is there a tutorial on how to do it? Please help me!

**Reply by ayerrocolom** [reply]
The file is called "Anura3D.sln" in version 2022 (in previous versions this file was called DynamicMPM.sln instead). Thanks for spotting that error in the tutorial manual, we will correct it.

**Reply by kuailebozai** [reply]
I have this problem when calculating, before it was possible to calculate, I do not know what is wrong。
Problem Description：
“E:\GID 3wei\3. 3weicellmodel-pile cap\bianpo22.A3D>"E:/GID 3wei/3. 3weicellmodel-pile cap/bianpo22.A3D\Anura3D_2021.exe" "E:/GID 3wei/3. 3weicellmodel-pile cap/bianpo22.A3D\bianpo22"
[ERROR] : (time step:      1) **Error opening file**: E:/GID 3wei/3. 3weicellmodel-pile cap/bianpo22.A3D\bianpo22.OUT
E:\GID 3wei\3. 3weicellmodel-pile cap\bianpo22.A3D>PAUSE”

**Reply by murphy-john** [reply]
Do you have read/write permissions to the "bianpo22.A3D" folder? Is the "bianpo22.OUT" file opened in another program?

**Comment by Zoe00WANG** [comment]
Hello！Anura3D is very interesting. I am studying it recently. 
I tried to run the ‘’Column Collapse‘’ example following  the tutorial manual, and  an error occurred (forrtl: error (65): floating invalid). 
Can you tell me how to solve the problem? Thank you very much for your help!
![0f260ed4d7b16a39ea203e4ec0e89f5](https://user-images.githubusercontent.com/112104448/186710978-5cfc23a8-116f-42bb-b48e-6c32f86b7ca7.jpg)

**Reply by murphy-john** [reply]
Will you confirm you've followed the column collapse example exactly as it's written in the tutorial manual? Or are you doing some modification?

Also, have you tried debugging the code in Visual Studio? The error is commonly due to a "dividing by zero" issue. Checking the input values and then debugging may help you identify where the issue comes from.

**Reply by Shachen75** [reply]
Hi, did you solve this floating invalid problem?

**Comment by QuantumNovice** [comment]
Where can I find prebuilt binaries?

**Reply by murphy-john** [reply]
We do not provide binaries. The code can be compiled as directed in the Tutorial Manual.

**Comment by BleachHwang** [comment]
I have followed the manual, and installed the VS Community version 2019 and the oneAPI package (Base + HPC Toolkits). However, I failed to compile the codes. "3 succeeded, 1 failed", as can be seen in the screenshot. Could you please tell me the potential causes? Thank you very much.
![image](https://user-images.githubusercontent.com/90396667/194982139-4ab31f14-72c2-4da1-bd27-8af7db81f5b6.png)

**Reply by BleachHwang** [reply]
The major error is "#7002: error in opening the compiled module file. Check INCLUDE paths". What should I do to deal with this error?

**Reply by BleachHwang** [reply]
![image](https://user-images.githubusercontent.com/90396667/194989191-5bc7bc24-0364-499a-8152-eb12dc3b7661.png)

**Reply by BleachHwang** [reply]
This problem is solved. I found a bug in the code. In Module ModEmptyElements (i.e., file MPMEmptyElements), Line 716, a redundant “&” 
is found. Please delete it then the codes can be compiled.
![image](https://user-images.githubusercontent.com/90396667/195025903-2d58230c-c035-42fc-9d32-3ab5d3d09156.png)

**Reply by ayerrocolom** [reply]
Thanks!

**Reply by EmreDumlu-UM** [reply]
Thank you!

**Reply by ReySonic** [reply]
Thank you.

**Comment by vjagodnik** [comment]
Hello,
I have more general question about compiling an exe file.
If compilation for version 2021 went flawlessly, is there a reason why 2023 version shows errors?
Also, I get four errors about missing "rc.exe". Unfortunaley i do not have access to the computer where I have VS, and on Mac i was unable to compile it.

**Reply by vjagodnik** [reply]
This is the error I get. 
Previous version of Anura was compiled without any error 
![image](https://github.com/Anura3D/Anura3D_OpenSource/assets/36809874/39a721b1-0be1-49fa-aeed-1f67bd0e17e4)

**Reply by vjagodnik** [reply]
Tried to install newer version of VS Community.
New errors.
![err_vs_1](https://github.com/Anura3D/Anura3D_OpenSource/assets/36809874/d34569f1-ce6d-4cd3-a5dd-acd31f01ac9e)

![err_vs_2](https://github.com/Anura3D/Anura3D_OpenSource/assets/36809874/5401ff72-3638-4241-9c8f-788e6d119045)

**Reply by vjagodnik** [reply]
Finally worked.
VS Community 2019  version 16.11.27 32bit did the job

**Reply by murphy-john** [reply]
Glad to hear the issue was resolved. I think it was caused by the your update of Visual Studio not recognizing the Intel oneAPI tools (which includes Fortran compiler)

**Comment by mozadst** [comment]
Hello, 
I've encountered the same problem as vjagodnik. I've been trying to compile ANURA3D for a while. I've tried installing and reinstalling different versions of VS 2022 and VS 2019 + Intel Toolkit and so far, I've only succeeded in building the file in VS 2019 16.11.53 
(Note: It says 1 succeeded, and theres quite a lot of warning about zlib). 
<img width="1920" height="1080" alt="anura3d build" src="https://github.com/user-attachments/assets/cf02ecae-54aa-4007-ae87-35819cf21555" />

Alas, when I tried to open the ANURA3D.exe, I've only greeted with the message of it missing zlib.dll
<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/dea72686-3c18-4bf1-8844-b81e255ef772" />

I would like an assistance for this problem because I desperately need this program for my undergraduate thesis. Thank you!
You can find me here: mozadst@gmail.com
